### ISSUE TITLE
Several inconsistencies in _loss_mmod

### ISSUE THREAD

**Issue by gregorkrajcovic** [davisking/dlib#470]
I analyzed loss_mmod_::compute_loss_value_and_gradient() and compared it with the original document https://arxiv.org/pdf/1502.00046.pdf. I found several potential bugs I would like to discuss.
**---- #1 ----**
In the document describing the algorithm to perform MMOD object detection on page 5, in the pseudo code of the algorithm on line 5 there is: 
_if D i does not overlap {D i−1 ,D i−2 ,...} then_
In c++ implementation there is:
               _for (unsigned long i = 0; i < dets.size() && final_dets.size() < max_num_dets; ++i)
                {
                    if (overlaps_any_box_nms(final_dets, dets[i].rect))
                        continue;
                    const std::pair<double,unsigned int> hittruth = find_best_match(*truth, dets[i].rect);
                    final_dets.push_back(dets[i].rect);_

Let's consider this case: set of detections contains 3 detections in following configuration: D1 overlaps D2, D2 overlaps D3 but D3 does not overlap D1.
Difference: **pseudocode algorithm would process only D1, C++ implementation would process D1 and D3.**

**---- #2 ----**
In c++ code there is a loop over all input samples (images): 
            _const float* out_data = output_tensor.host();
            for (long i = 0; i < output_tensor.num_samples(); ++i)
            {
                tensor_to_dets(input_tensor, output_tensor, i, dets, -options.loss_per_false_alarm, sub);
                .....
                for (auto&& x : *truth)
                {
                    if (!x.ignore)
                    {
                        point p = image_rect_to_feat_coord(input_tensor, x, sub);
                        loss -= out_data[p.y()*output_tensor.nc() + p.x()];_
  
While method **tensor_to_dets** takes into account index i of the processed image and adjusts pointer to raw data like this: _const float* out_data = output_tensor.host() + output_tensor.nr()*output_tensor.nc()*i;_
it is forgotten in the loss calculation here:  _loss -= out_data[p.y()*output_tensor.nc() + p.x()];_
The same problem occurs at the end of the cycle:
               _for (auto&& x : final_dets)
                {
                    loss += out_data[x.tensor_offset];_
Here _tensor_offset_ again does not contain the image index adjustment. 
**In both cases output data of the 1st image is used in the particular calculation for all of the images.

**---- #3 ----**
Intersection over union is implemented strangely in c++. Here is the code:
   _inline double box_intersection_over_union (
        const drectangle& a,
        const drectangle& b
    ) 
    {
        const double inner = a.intersect(b).area();
        if (inner == 0)
            return 0;
        **const double outer = (a+b).area();**
        return inner/outer;
    }_

Outer area is calculated for the rectangle calculated from a+b operation. This operation returns rectangle bounding both input rectangles. Area of bounding rectangle can be significantly larger than area of an union of rectangles. Correct implementation should be:
outer = a.area()+b.area()-inner;

**Comment by davisking** [comment]
The C++ code is right.  For the first thing, yeah, that line in the paper
is confusing and somewhat wrong.  It should say you only skip rectangles
that overlap with a "detected" rectangle so that the behavior matches that
in algorithm 1, which is very important.  Then for the second issue, you
are missing the part of the code that adjusts out_data appropriately.

**Comment by gregorkrajcovic** [comment]
Aha, I see _out_data += output_tensor.nr()*output_tensor.nc();_ at the end of the cycle. My mistake. Is there any reason, why not to do _loss += x.detection_confidence_  instead of _out_data[x.tensor_offset]_ ?
What about intersection over union - should it not be _inner/(a.area()+b.area()-inner)_ ?

**Comment by gregorkrajcovic** [comment]
And another question: is there any unspoken requirement, that the last layer must operate in negative range ? 
Why do I ask: in this step: _tensor_to_dets(input_tensor, output_tensor, i, dets, -options.loss_per_false_alarm, sub);_ negative threshold is used to get potential detections. If the output tensor is never positive or zero, than all of its points are always considered as detections. Maybe it does not spoil the detection it self, but it has huge performance impact.

**Comment by davisking** [comment]
That's just not how the math works out. 

Yeah maybe I should change that to calculate it in the usual way.  I don't think it matters as far as algorithm results though.  But it would be less confusing to users.

If the output tensor is never positive then there aren't any detections, except in the loss augmented detection part.  Then there are only detections if the output is >= -1 by default.
